File: app/models/onenotify_bb.py
from sqlalchemy import Boolean, Column, DateTime, Float, ForeignKey, Integer, JSON, String, Text
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func

from app.db.session import Base


ONB_STATUS_RECEBIDA = "RECEBIDA"
ONB_STATUS_CONCILIADA_AUTO = "CONCILIADA_AUTO"
ONB_STATUS_PENDENTE_FLOW = "PENDENTE_FLOW"
ONB_STATUS_PENDENTE_DOCUMENTO = "PENDENTE_DOCUMENTO"
ONB_STATUS_REVISAO = "REVISAO"
ONB_STATUS_ERRO = "ERRO"

ONB_ACTION_SEM_TRATAMENTO_NOTIFY = "SEM_TRATAMENTO_NOTIFY"
ONB_ACTION_TRATAR_DOCUMENTO_FLOW = "TRATAR_DOCUMENTO_FLOW"
ONB_ACTION_REVISAR_MANUALMENTE = "REVISAR_MANUALMENTE"


class OneNotifyBBNotification(Base):
    __tablename__ = "onenotify_bb_notificacoes"

    id = Column(Integer, primary_key=True, index=True)
    external_group_id = Column(String, nullable=False, unique=True, index=True)
    source = Column(String, nullable=False, default="ONENOTIFY_BB", index=True)
    schema_version = Column(String, nullable=True)

    notify_ids = Column(JSON, nullable=True)
    npj = Column(String, nullable=True, index=True)
    data_notificacao = Column(String, nullable=True, index=True)
    notification_date_iso = Column(String, nullable=True, index=True)
    publication_date = Column(String, nullable=True, index=True)

    numero_processo_cnj = Column(String, nullable=True, index=True)
    cnj_publicacao = Column(String, nullable=True, index=True)
    cnj_principal_notify = Column(String, nullable=True, index=True)
    cnj_divergent = Column(Boolean, nullable=False, default=False, index=True)

    adverso_principal = Column(String, nullable=True)
    polo = Column(String, nullable=True)
    posicao_cliente = Column(String, nullable=True, index=True)
    tipos_notificacao = Column(JSON, nullable=True)

    rpa_status = Column(JSON, nullable=True)
    bb_ciencia_status = Column(JSON, nullable=True)
    human_status = Column(JSON, nullable=True)
    flow_sync_status = Column(JSON, nullable=True)
    status_legacy = Column(JSON, nullable=True)

    flow_status = Column(String, nullable=False, default=ONB_STATUS_RECEBIDA, index=True)
    action_suggested = Column(String, nullable=True, index=True)
    match_strategy = Column(String, nullable=True)
    match_score = Column(Float, nullable=True, index=True)
    match_reason = Column(Text, nullable=True)

    matched_publication_record_id = Column(
        Integer,
        ForeignKey("publicacao_registros.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )
    matched_legal_one_update_id = Column(Integer, nullable=True, index=True)
    matched_publication_status = Column(String, nullable=True, index=True)

    andamentos = Column(JSON, nullable=True)
    documentos = Column(JSON, nullable=True)
    conteudo = Column(JSON, nullable=True)
    raw_payload = Column(JSON, nullable=True)
    text_content = Column(Text, nullable=True)
    document_summary = Column(JSON, nullable=True)

    last_error = Column(Text, nullable=True)
    received_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    reconciled_at = Column(DateTime(timezone=True), nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), onupdate=func.now(), nullable=True)

    matched_publication = relationship("PublicationRecord")

File: frontend/src/services/onenotify-bb.ts
import { apiFetch } from "@/lib/api-client";

const BASE = "/api/v1/publications/onenotify-bb";

async function json<T>(response: Response): Promise<T> {
  if (!response.ok) {
    const text = await response.text();
    throw new Error(text || `Erro HTTP ${response.status}`);
  }
  return response.json();
}

export interface OneNotifyBBStats {
  total: number;
  matched: number;
  matched_pct: number;
  auto_conciliated: number;
  auto_conciliated_pct: number;
  cnj_divergent: number;
  pending_document: number;
  no_match: number;
}

export interface OneNotifyBBSummary {
  id: number;
  external_group_id: string;
  source: string;
  notify_ids: number[];
  npj?: string;
  data_notificacao?: string;
  notification_date_iso?: string;
  publication_date?: string;
  numero_processo_cnj?: string;
  cnj_publicacao?: string;
  cnj_principal_notify?: string;
  cnj_divergent: boolean;
  adverso_principal?: string;
  posicao_cliente?: string;
  tipos_notificacao: string[];
  flow_status: string;
  action_suggested?: string;
  matched_publication_record_id?: number;
  matched_legal_one_update_id?: number;
  matched_publication_status?: string;
  match_score?: number;
  match_strategy?: string;
  document_summary?: Record<string, unknown>;
}

export interface DiffRow {
  kind: "equal" | "replace" | "delete" | "insert";
  left_line?: number | null;
  right_line?: number | null;
  left: string;
  right: string;
}

export interface OneNotifyBBDetail extends OneNotifyBBSummary {
  andamentos: unknown[];
  documentos?: Record<string, unknown> | null;
  conteudo: Record<string, unknown>;
  raw_payload: Record<string, unknown>;
  text_content: string;
  match_reason?: string;
  matched_publication?: Record<string, unknown> | null;
  diff: {
    score: number;
    rows: DiffRow[];
  };
}

export async function getOneNotifyBBStats(): Promise<OneNotifyBBStats> {
  return json(await apiFetch(`${BASE}/stats`));
}

export async function listOneNotifyBBNotifications(params: {
  status?: string;
  action?: string;
  q?: string;
  limit?: number;
  offset?: number;
}): Promise<{ total: number; limit: number; offset: number; items: OneNotifyBBSummary[] }> {
  const qs = new URLSearchParams();
  if (params.status) qs.set("status", params.status);
  if (params.action) qs.set("action", params.action);
  if (params.q) qs.set("q", params.q);
  qs.set("limit", String(params.limit ?? 50));
  qs.set("offset", String(params.offset ?? 0));
  return json(await apiFetch(`${BASE}/notifications?${qs.toString()}`));
}

export async function getOneNotifyBBDetail(id: number): Promise<OneNotifyBBDetail> {
  return json(await apiFetch(`${BASE}/notifications/${id}`));
}
